File: DS_and_A/cpp/mergesort.cpp
#include <iostream>
#include <vector>

using namespace std;

// Merge function to merge two halves of the array
void merge(vector<int>& array, int left, int middle, int right) {
    int leftSize = middle - left + 1;
    int rightSize = right - middle;

    // Create temporary arrays
    vector<int> leftArray(leftSize), rightArray(rightSize);

    // Copy data to temporary arrays leftArray[] and rightArray[]
    for (int i = 0; i < leftSize; i++)
        leftArray[i] = array[left + i];
    for (int j = 0; j < rightSize; j++)
        rightArray[j] = array[middle + 1 + j];

    // Merge the temporary arrays back into array[left..right]
    int i = 0; // Initial index for leftArray[]
    int j = 0; // Initial index for rightArray[]
    int k = left; // Initial index for merged subarray
    while (i < leftSize && j < rightSize) {
        if (leftArray[i] <= rightArray[j]) {
            array[k] = leftArray[i];
            i++;
        } else {
            array[k] = rightArray[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of leftArray[], if any
    while (i < leftSize) {
        array[k] = leftArray[i];
        i++;
        k++;
    }

    // Copy the remaining elements of rightArray[], if any
    while (j < rightSize) {
        array[k] = rightArray[j];
        j++;
        k++;
    }
}

// Merge sort function
void mergeSort(vector<int>& array, int left, int right) {
    if (left < right) {
        // Find the middle point
        int middle = left + (right - left) / 2;

        // Sort first and second halves
        mergeSort(array, left, middle);
        mergeSort(array, middle + 1, right);

        // Merge the sorted halves
        merge(array, left, middle, right);
    }
}

// Utility function to print an array
void printArray(const vector<int>& array) {
    for (int num : array)
        cout << num << " ";
    cout << endl;
}

// Main function
int main() {
    vector<int> array = {12, 11, 13, 5, 6, 7};
    int size = array.size();

    cout << "Given array is \n";
    printArray(array);

    mergeSort(array, 0, size - 1);

    cout << "\nSorted array is \n";
    printArray(array);

    return 0;
}

File: DS_and_A/recursion_basics/merge_sort.java
public class merge_sort {
    
    private static void mergesort(int[] arr){
        int length = arr.length;
        int middle = length/2;
        if(length<=1)return;
        int[] leftarray = new int[middle];
        int[] rightarray = new int[length-middle];
        int i =0; //leftarray pointer
        int j=0; //rightarray pointer
        for(;i<length;i++){
            if(i<middle){
                leftarray[i]=arr[i];
            }
            else{
                rightarray[j]=arr[i];
                j++;
            }
        }
        mergesort(leftarray);
        mergesort(rightarray);
        merge(leftarray, rightarray, arr);

    }
    private static void merge(int[] leftarray ,int[] rightarray,int[] array ){
        int leftSize = array.length/2;
        int rightSize=array.length-leftSize;
        int i=0;
        int l=0;
        int r =0;
        while (l<leftSize && r<rightSize) {
            if(leftarray[l]<rightarray[r]){
                array[i]=leftarray[l];
                i++;
                l++;
            }
            else{
                array[i]=rightarray[r];
                r++;
                i++;
            }
        }
        while (l<leftSize) {
            array[i]=leftarray[l];
            i++;
            l++;
        }
        while (r<rightSize) {
            array[i]=rightarray[r];
            i++;
            r++;
        }
    }
    public static void main(String[] args) {
        int[] a = {9,8,7,6,5,4,3,2,1};
        mergesort(a);
        for(int i=0;i<a.length;i++){
            System.out.println(a[i]);
        }
    }
}
